internal/service/advancedclustertpf/data_source.go

```diff
@@ -38,10 +41,88 @@ func (d *ds) Schema(ctx context.Context, req datasource.SchemaRequest, resp *dat
 }
 
 func (d *ds) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	var state TFModelDS
+	diags := &resp.Diagnostics
+	diags.Append(req.Config.Get(ctx, &state)...)
+	if diags.HasError() {
+		return
+	}
+	model := d.readCluster(ctx, diags, &state)
+	if model != nil {
+		diags.Append(resp.State.Set(ctx, model)...)
+	}
+}
+
+func (d *ds) readCluster(ctx context.Context, diags *diag.Diagnostics, modelDS *TFModelDS) *TFModelDS {
+	clusterName := modelDS.Name.ValueString()
+	projectID := modelDS.ProjectID.ValueString()
+	useReplicationSpecPerShard := modelDS.UseReplicationSpecPerShard.ValueBool()
+	api := d.Client.AtlasV2.ClustersApi
+	clusterResp, _, err := api.GetCluster(ctx, projectID, clusterName).Execute()
+	if err != nil {
+		if admin.IsErrorCode(err, ErrorCodeClusterNotFound) {
+			return nil
+		}
+		diags.AddError("errorRead", fmt.Sprintf(errorRead, clusterName, err.Error()))
+		return nil
+	}
+	modelIn := &TFModel{
+		ProjectID: modelDS.ProjectID,
+		Name:      modelDS.Name,
+	}
+	// TODO: pass !UseReplicationSpecPerShard to overrideUsingLegacySchema
+	modelOut, extraInfo := getBasicClusterModel(ctx, diags, d.Client, clusterResp, modelIn)
+	if diags.HasError() {
+		return nil
+	}
+	if extraInfo.AsymmetricShardUnsupportedError && !useReplicationSpecPerShard {
+		diags.AddError("errorRead", "Please add `use_replication_spec_per_shard = true` to your data source configuration to enable asymmetric shard support. Refer to documentation for more details.")
+		return nil
+	}
+	updateModelAdvancedConfig(ctx, diags, d.Client, modelOut, nil, nil)
+	if diags.HasError() {
+		return nil
+	}
+	modelOutDS, err := conversion.CopyModel[TFModelDS](modelOut)
+	if err != nil {
+		diags.AddError(errorRead, fmt.Sprintf("error setting model: %s", err.Error()))
+		return nil
+	}
+	modelOutDS.UseReplicationSpecPerShard = modelDS.UseReplicationSpecPerShard // attrs not in resource model
+	return modelOutDS
 }
 
-// TODO: see if resource model can be used instead, probably different only in timeouts
-type ModelDS struct {
-	ProjectID types.String `tfsdk:"project_id"`
-	Name      types.String `tfsdk:"name"`
+// TODO: difference with TFModel: misses timeouts, adds use_replication_spec_per_shard.
+type TFModelDS struct {
```

Comment:
WDYT about moving this and TFModel to be next to each other? 
I suspect the models will change together, and having them next to each other can help avoid inconsistencies

Reply:
good idea, moved together and renamed files resource_schema to just schema in: https://github.com/mongodb/terraform-provider-mongodbatlas/pull/2865/commits/69909257748e9485841b1eabfa88824ab8488a36

Reply:
also moved schemas together: https://github.com/mongodb/terraform-provider-mongodbatlas/pull/2865/commits/36973eb9828a557a1092a1a5c0ba681f4c0e066e